File: memoization/memoization.js
function memoize(f) {
  const cache = new Map();
  return function (...args) {
    const key = JSON.stringify(args);
    if (cache.has(key)) {
      return cache.get(key);
    } else {
      const result = f(...args);
      cache.set(key, result);
      return result;
    }
  }
}

function slowFunction(x) {
  console.log('Computing...');
  return x * x;
}

const memoizedFunction = memoize(slowFunction);

console.log(memoizedFunction(5)); // Logs "Computing..." and then 25
console.log(memoizedFunction(5)); // Logs 25 (retrieved from cache, no "Computing...")
console.log(memoizedFunction(6)); // Logs "Computing..." and then 36

File: memoization/memoization.py
def memoize(func):
    cache = dict()

    def memoized_func(*args):
        if args in cache:
            return cache[args]
        result = func(*args)
        cache[args] = result
        return result

    return memoized_func

def slowFunction(x):
  print('Computing...')
  return x * x

memoizedFunction = memoize(slowFunction)

print(memoizedFunction(5)); # Logs "Computing..." and then 25
print(memoizedFunction(5)); # Logs 25 (retrieved from cache, no "Computing...")
print(memoizedFunction(6)); # Logs "Computing..." and then 36